WordPress/Classes/ViewRelated/Menus/MenuItemEditingViewController.m

```diff
@@ -0,0 +1,470 @@
+{
+    self.sourceViewIsTyping = YES;
+    [self updateLayoutIfNeededAnimated];
+}
+
+- (void)sourceViewControllerDidEndEditingWithKeyboard:(MenuItemSourceViewController *)sourceViewController
+{
+    self.sourceViewIsTyping = NO;
+    [self updateLayoutIfNeededAnimated];
+}
+
+#pragma mark - MenuItemEditingHeaderViewDelegate
+
+- (void)editingHeaderView:(MenuItemEditingHeaderView *)headerView didUpdateTextForItemName:(NSString *)text
+{
+    self.item.name = text.length ? text : nil;
+    [self.sourceViewController refreshForUpdatedItemName];
+}
+
+#pragma mark - MenuItemEditingFooterViewDelegate
+
+- (void)editingFooterViewDidSelectSave:(MenuItemEditingFooterView *)footerView
+{
+    // Update the original item in the correct context with any changes made in the scratch context/item.
+    NSDictionary *changesValues = [self.item changedValues];
+    if (changesValues.count > 0) {
+        MenuItem *itemInMainContext = [self.blog.managedObjectContext objectRegisteredForID:self.item.objectID];
+        [itemInMainContext setValuesForKeysWithDictionary:changesValues];
+    }
+    if (self.onSelectedToSave) {
+        self.onSelectedToSave();
+    }
+}
+
+- (void)editingFooterViewDidSelectTrash:(MenuItemEditingFooterView *)footerView
+{
+    if (self.onSelectedToTrash) {
+        self.onSelectedToTrash();
+    }
+}
+
+- (void)editingFooterViewDidSelectCancel:(MenuItemEditingFooterView *)footerView
+{
+    if (self.onSelectedToCancel) {
+        self.onSelectedToCancel();
+    }
+}
+
+#pragma mark - notifications
+
+- (void)updateWithKeyboardNotification:(NSNotification *)notification
+{
+    CGRect frame = [[notification.userInfo objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
+    frame = [self.view.window convertRect:frame toView:self.view];
+    
+    CGFloat constraintConstant = self.stackViewBottomConstraint.constant;
+    
+    if (frame.origin.y > self.view.frame.size.height) {
+        constraintConstant = 0.0;
+    } else  {
+        constraintConstant = self.view.frame.size.height - frame.origin.y;
+    }
+    
+    [self.view layoutIfNeeded];
+    self.stackViewBottomConstraint.constant = constraintConstant;
+    
+    NSTimeInterval duration = [[notification.userInfo objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
+    UIViewAnimationOptions options = [[notification.userInfo objectForKey:UIKeyboardAnimationCurveUserInfoKey] integerValue];
+    [UIView animateWithDuration:duration delay:0.0 options:options animations:^{
+        
+        [self.view layoutIfNeeded];
+        
+    } completion:nil];
+}
+
+- (void)keyboardWillHideNotification:(NSNotification *)notification
+{
+    self.observesKeyboardChanges = NO;
+    self.stackViewBottomConstraint.constant = 0;
+    [self.view layoutIfNeeded];
+}
+
+- (void)keyboardWillShowNotification:(NSNotification *)notification
+{
+    self.observesKeyboardChanges = YES;
```

Comment:
`observesKeyboardChanges` seems a misleading name since the vc is *always* observing keyboard changes. Its really a flag for whether or not to take action in response to a change. 
Maybe it could be renamed to clarify its purpose?  Alternatively, perhaps *this* is where an observer should be registered for keyboard frame change notifications, and then unregistered in the will-hide handler?